talisker/testing.py

```diff
@@ -0,0 +1,273 @@
+        except ValueError:
+            assert 0, "failed to parse logfmt: " + log
+        return {
+            'ts': date + " " + time,
+            'level': level,
+            'logger': name,
+            'logmsg': msg,
+            'extra': extra,
+            'trailer': trailer,
+        }
+
+    def _compare_strings(self, needle, haystack):
+        return (
+            all(k in haystack for k in needle) and
+            all(needle[k] in haystack[k] for k in needle)
+        )
+
+    def exists(self, **match):
+        extra = match.pop('extra', None)
+        trailer = match.pop('trailer', None)
+        for log in self.logs:
+            strings_match = self._compare_strings(match, log)
+            extra_match = trailer_match = True
+
+            if extra is not None:
+                extra_match = self._compare_strings(extra, log['extra'])
+
+            if trailer is not None:
+                trailer_match = True
+                for line in trailer:
+                    if not any(line in tline for tline in log['trailer']):
+                        trailer_match = False
+                        break
+
+            if strings_match and extra_match and trailer_match:
+                return True
+
+        return False
+
+    def __contains__(self, params):
+        return self.exists(**params)
+
+
+class ServerProcessError(Exception):
+    pass
+
+
+class ServerProcess(object):
+    """Context mananger to run a server subprocess """
+    def __init__(self, cmd):
+        self.cmd = cmd
+        self.output = []
+        self.ps = None
+        self._log = None
+
+    @property
+    def finished(self):
+        return self.ps is not None and self.ps.poll() is not None
+
+    @property
+    def log(self):
+        assert self.finished
+        if self._log is None:
+            self._log = LogOutput(self.output)
+        return self._log
+
+    def check(self):
+        """Check the server process is still running or has finished."""
+        rc = self.ps.poll()
+        if rc is not None and rc > 0:
+            raise ServerProcessError('subprocess errored')
+
+    def handle_exception(self, type=None, value=None, traceback=None):
+        """Clean up and log process exception, without supressing it."""
+        if not self.finished:
+            self.ps.terminate()
+
+        # ensure all output is read
+        self.output.extend(self.ps.stdout)
+
+        # just dump the output to stderr for now, for visibility
+        # might be a way to include it in the exception somehow
+        if type is ServerProcessError:
+            sys.stderr.write('Server process died:\n')
+            sys.stderr.write(''.join(self.output))
+
+    @contextmanager
+    def _handle_enter_exception(self):
+        """Helper to manually handle an exception in the __enter__ method"""
+        try:
+            yield
+        except:
+            suppressed = self.handle_exception(*sys.exc_info())
```

Comment:
`handle_exception` always returns None, so the conditional on the next line is not meaningful. I think an unexpected server process exit should be supressed?

Reply:
I added it incase future code did return something in handle_exception, so that supression would work with in this odd 'chained' case. But I've reworked it now, so it's not needed.